### ISSUE TITLE
invalid_challenge: "Correct value not found for DNS challenge"

### ISSUE THREAD

**Issue by qnxor** [kappataumu/letsencrypt-cloudflare-hook#27]
I installed all prerequisites as mentioned in the README and then exported the following

```
export CF_EMAIL=<MY_CF_EMAIL>
export CF_KEY=<MY_CF_API_KEY>
export CF_DEBUG=true
export CF_DNS_SERVERS='8.8.8.8 8.8.4.4'
```

Then executed the following below which eventually failed. I note that I could see the new TXT records were in fact generated and deleted (I was refreshing my CloudFlare dashboard and was seeing them appear) ... but for some reason the challenge failed. Is it a DNS propagation issue? Any other clues?

```
# dehydrated/dehydrated -c -t dns-01 -k -d DOMAIN.TLD -d mail.DOMAIN.TLD -d www.DOMAIN.TLD dehydrated/hooks/cloudflare/hook.py
# INFO: Using main config file /root/dehydrated/config
Processing DOMAIN.TLD with alternative names: mail.DOMAIN.TLD www.DOMAIN.TLD
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting challenge for DOMAIN.TLD...
 + Requesting challenge for mail.DOMAIN.TLD...
 + Requesting challenge for www.DOMAIN.TLD...
 + CloudFlare hook executing: deploy_challenge
 + Creating TXT record: DOMAIN.TLD => kIcGIJTbYgjfXoLrfEBfsrFj4VjtQMfEehTjKAZMNWM
 + Unable to locate record named _acme-challenge.DOMAIN.TLD
 + TXT record created, ID: b3b68cead9fed0252a2cf327b15941fc
 + Settling down for 10s...
 + Responding to challenge for DOMAIN.TLD...
 + CloudFlare hook executing: clean_challenge
 + Deleting TXT record name: _acme-challenge.DOMAIN.TLD
 + CloudFlare hook executing: invalid_challenge
 + CloudFlare hook executing: deploy_challenge
 + Creating TXT record: mail.DOMAIN.TLD => WY37eSk0fYz8HOXfLo4PlPZ3fYdh6RUfEnqR484-4_Y
 + Unable to locate record named _acme-challenge.mail.DOMAIN.TLD
 + TXT record created, ID: 8b642ad0cef522c9e633439b5e50a602
 + Settling down for 10s...
 + Responding to challenge for mail.DOMAIN.TLD...
 + CloudFlare hook executing: clean_challenge
 + Deleting TXT record name: _acme-challenge.mail.DOMAIN.TLD
 + CloudFlare hook executing: invalid_challenge
 + CloudFlare hook executing: deploy_challenge
 + Creating TXT record: www.DOMAIN.TLD => pVVB-_LekLUCcXNx-k1XSNUIgTcrVQrfGBZOlloHEFM
 + Unable to locate record named _acme-challenge.www.DOMAIN.TLD
 + TXT record created, ID: 6976f699bde7a05ee6a93fc1e0d02c68
 + Settling down for 10s...
 + Responding to challenge for www.DOMAIN.TLD...
 + CloudFlare hook executing: clean_challenge
 + Deleting TXT record name: _acme-challenge.www.DOMAIN.TLD
 + CloudFlare hook executing: invalid_challenge
ERROR: Challenge is invalid! (returned: invalid) (result: {
  "type": "dns-01",
  "status": "invalid",
  "error": {
    "type": "urn:acme:error:unauthorized",
    "detail": "Correct value not found for DNS challenge",
    "status": 403
  },
  "uri": "https://acme-v01.api.letsencrypt.org/acme/challenge/5rkcG4LZEKNo_CDlkUHr0gjs1BrhJp5CbbsJjE2y4ws/446249521",
  "token": "5qRUM02cxrTBBusrd-S4nKjczrFz8p3-k6dvbZ3Adbw",
  "keyAuthorization": "5qRUM02cxrTBBusrd-S4nKjczrFz8p3-k6dvbZ3Adbw.B5SOsPgZwhlOngI1kKY1iLN_1jkjJ-Y7idMO1bvsbJE"
})
```

**Comment by kappataumu** [comment]
Hi, what happens if you only specify a single domain? Could you post the output?

**Comment by qnxor** [comment]
Same error, but let me debug a few more things first as I recently changed my nameservers ... while your script uses Google's DNS server (8.8.8.8) to query my DNS records (which does see the new nameservers), the Let's Encrypt server is most likely using a different server to query my DNS records and may not see the changes ...

**Comment by kappataumu** [comment]
If you share the output for the single domain run, using `export CF_DEBUG=true`, maybe I could offer more help.

**Comment by qnxor** [comment]
It was literally the same error, see below (as I said, I think the issue lies with the usage of different servers to check the DNS entries, which see different things due to propagation delays)

```
root@xylem:~# dehydrated/dehydrated -c -t dns-01 -d DOMAIN.TLD -k dehydrated/hooks/cloudflare/hook.py
# INFO: Using main config file /root/dehydrated/config
Processing DOMAIN.TLD
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting challenge for DOMAIN.TLD...
 + CloudFlare hook executing: deploy_challenge
 + Creating TXT record: DOMAIN.TLD => qcZDySC1swnNFtL4wgpsy4my4Ku7VSWkBvnNyG2Sccw
 + Unable to locate record named _acme-challenge.DOMAIN.TLD
 + TXT record created, ID: 2273343e891575303e06670ab4ac7d71
 + Settling down for 10s...
 + Responding to challenge for DOMAIN.TLD...
 + CloudFlare hook executing: clean_challenge
 + Deleting TXT record name: _acme-challenge.DOMAIN.TLD
 + CloudFlare hook executing: invalid_challenge
ERROR: Challenge is invalid! (returned: invalid) (result: {
  "type": "dns-01",
  "status": "invalid",
  "error": {
    "type": "urn:acme:error:unauthorized",
    "detail": "Correct value not found for DNS challenge",
    "status": 403
  },
  "uri": "https://acme-v01.api.letsencrypt.org/acme/challenge/x4RQw-QqM8nILjmaFokIetRgtq_qQsg9BfOtk6NKC7k/446449610",
  "token": "8Ip3_DWPA6siQVDGzmGYu91O71mt-T5xxRY09T87xBQ",
  "keyAuthorization": "8Ip3_DWPA6siQVDGzmGYu91O71mt-T5xxRY09T87xBQ.B5SOsPgZwhlOngI1kKY1iLN_1jkjJ-Y7idMO1bvsbJE"
})
```

**Comment by kappataumu** [comment]
Yeah, the single run probably wasn't needed in this case, but the chained hook call output was a bit messy. Will push an update with more helpful debug information shortly. 

Since you can see the TXT records being created in Cloudflare (with correct values), it may indeed come down to propagation issues. Do report back with anything new.

**Comment by denniseijpe** [comment]
This is probably fixed by PR #29. I had the same error message.

The log also shows the hook isn't waiting for the DNS to update:

```
 + Settling down for 10s...
 + Responding to challenge for DOMAIN.TLD...
 + CloudFlare hook executing: clean_challenge
```

**Comment by kappataumu** [comment]
OK, I'm closing this for now, @qnxor feel free to post if you are still having an issue.
